functions/go/upsert-resource/upsertresource/upsert_resource.go

```diff
@@ -0,0 +1,148 @@
+package upsertresource
+
+import (
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/fn/runtime/runtimeutil"
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// UpsertResource upserts resource to input list of resources
+// resources are uniquely identifies by Group, Kind, Name and Namespace
+type UpsertResource struct {
+	// Resource is the input resource for upsert
+	Resource *yaml.RNode
+}
+
+// Filter implements UpsertResource as a yaml.Filter
+func (ur UpsertResource) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	replaced, err := ReplaceResource(nodes, ur.Resource)
+	if err != nil {
+		return nodes, err
+	}
+	if !replaced {
+		return AddResource(nodes, ur.Resource)
+	}
+	return nodes, nil
+}
+
+// ReplaceResource checks if the inputResource matches with any of the resources in nodes list
+// and replaces it with inputResource
+// nodes are matched based on Group, Kind, Name and Namespace
+func ReplaceResource(nodes []*yaml.RNode, inputResource *yaml.RNode) (bool, error) {
+	found := false
+	inputMeta, err := inputResource.GetMeta()
+	if err != nil {
+		return false, err
+	}
+	for i := range nodes {
+		rMeta, err := nodes[i].GetMeta()
+		if err != nil {
+			return false, err
+		}
+		// skip processing resource if it is a function config
+		if IsFunctionConfig(rMeta) {
```

Comment:
FYI In kpt 1.0, you can also operate on a meta resource  (controlled by the --include-meta-resource flag)

Reply:
Yes. But this annotation only exists in v0.X. So this check is to only filter resources in v0.X. Added TODO to clean this up after v0.X deprecation.